libsupport/include/katana/ProgressTracer.h

```diff
@@ -0,0 +1,222 @@
+#ifndef KATANA_LIBSUPPORT_KATANA_PROGRESSTRACER_H_
+#define KATANA_LIBSUPPORT_KATANA_PROGRESSTRACER_H_
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "katana/Result.h"
+#include "katana/config.h"
+
+/// This tracer does not currently support thread-local tracers or concurrency controls
+/// Functions should only be used in a single-threaded context
+///
+/// Best Practices:
+///   If possible always avoid creating raw ProgressSpans from ProgressTracer
+///   If possible always use ProgressScopes to handle ProgressSpans
```

Comment:
What is the difference b/w tracer, scope, and span?

Reply:
These are from the OpenTracing specifications which can be found here: https://opentracing.io/docs/overview/spans/

In short spans are units of work with a defined start and stop point which we can add timestamped logs to as well as tags for the overall span (or unit of work)

Scopes are used to manage spans by being immovable and not copyable, and closing on RAII mechanics.  Whenever possible raw spans should not be used and scopes instead

Tracers are used to control span logic, they create spans and maintain the active span for ease of use for the developer

Reply:
Each Scope is "in charge of" a span for its lifetime, and closes it when it goes out of scope or the developer calls Close on it

Reply:
Thanks Patrick. Please add this into the doxygen comments before you describe best practices.